### PR TITLE
Upgrade test CLI tool

### PR DESCRIPTION
**Pull request by DaedalusG**
This PR introduces the upgrade test tool which is seeing wide use currently in debugging problems in the v5.3.0 release.

Currently its also the best tool for a quick feedback loop on changes to migrator.

The tool is capable of testing post and prerelease versions of migrator via flag options. Currently the `bazel test` endpoint isn't ready and specific targets will need to be written to plug in for CI testing. 

Moving to get this merged simply on the basis that its seeing wide use. Theres still some TODO notes etc and areas that can use refactoring but this is a pretty good MVP.

## Test plan

Built and run locally

`VERSION=5.3.1 bazel run //testing/tools/upgradetest:sh_upgradetest_run --config darwin-docker --stamp --workspace_status_command=./dev/bazel_stamp_vars.sh -- all`
^^ this is the current full test suite minus the `-mr` max routines flag to control goroutines

### CODE REVIEW

**Comment by jhchabran** [comment]
@DaedalusG happy to review this, but you need to fix the issue with the logging library here :)

**Comment by DaedalusG** [comment]
@jhchabran I fixed it! Although some upstream changes appear to have broken the stamp functionality:
```
➜  sourcegraph git:(wg/upgrade-test-tool-post-v5.3.0) ✗ VERSION=5.3.3 bazel run //testing/tools/upgradetest:sh_upgradetest_run --stamp --workspace_status_command=./dev/bazel_stamp_vars.sh -- auto -avs 5.2.7
WARNING: Build option --stamp has changed, discarding analysis cache (this can be expensive, see https://bazel.build/advanced/performance/iteration-speed).
INFO: Analyzed target //testing/tools/upgradetest:sh_upgradetest_run (0 packages loaded, 74102 targets configured).
ERROR: /Users/warrengifford/sourcegraph/cmd/migrator/airgappedgen/BUILD.bazel:28:8: Executing genrule //cmd/migrator/airgappedgen:tar_schema_descriptions failed: (Exit 1): bash failed: error executing Genrule command (from target //cmd/migrator/airgappedgen:tar_schema_descriptions) /bin/bash -c ... (remaining 1 argument skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
🔴 GH_TOKEN is not set. Please ensure it's set and accessible to Bazel.
🔴 If you're running this locally, append --action_env=GH_TOKEN to make it visible to Bazel.
🔴 If you see this in CI, please reach out on #discuss-dev-infra.
INFO: Found 1 target...
Target //testing/tools/upgradetest:sh_upgradetest_run failed to build
Aspect @@rules_rust//rust/private:clippy.bzl%rust_clippy_aspect of //testing/tools/upgradetest:sh_upgradetest_run up-to-date (nothing to build)
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 26.893s, Critical Path: 15.75s
INFO: 9 processes: 2 internal, 7 darwin-sandbox.
ERROR: Build did NOT complete successfully
ERROR: Build failed. Not running target
Error: bazel exited with exit code: 1
```

**Comment by jhchabran** [comment]
Awesome, I'll review it tomorrow morning :)

File: testing/tools/upgradetest/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+## Overview
+
+The upgradetest folder contains the code for release tooling intended used to asses the stability of a new migrator and frontend build before tagging a new release.
+The tooling takes the form of a cli interface and is intended to orchestrate a psuedo-release and upgrade using bazel built images. The tests are inocations of our services relevant to schema migrations and versioning.
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:4]:
```suggestion
The tooling takes the form of a cli interface and is intended to orchestrate the upgrade process using bazel built images. The tests are invocations of our services relevant to schema migrations and versioning.
```

File: testing/tools/upgradetest/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+This test does not test Sourcegraph features, only the basic operations of database and frontend versioning.
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:5]:
```suggestion
This test does not test Sourcegraph features, only the basic operations of database and frontend version upgrades and correctness checks.
```

File: testing/tools/upgradetest/README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+Commands are intended to be invoked using `bazil targets`, with each command executing tests intrended to be run in CI. The CI tests treat the current repo branch of `sourcegraph/sourcegraph` as a prospective release.
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:7]:
```suggestion
Commands are intended to be invoked through Bazel, with each command executing tests intended to be run in CI. The CI tests treat the current repo branch of `sourcegraph/sourcegraph` as a prospective release.
```

File: testing/tools/upgradetest/README.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+This "release branch" may be stamped with a version, or will be versioned `0.0.0+dev`. A stamped version of the release branch must be invoked with a certain bazel flags `--stamp`, `--workspace_status_command=./dev/bazel_stamp_vars.sh`,
+and additionally requires a `VERSION` env var to be set with a semantic version string `X.X.X`.
+
+The general idea of the tests is to verify that a given upgrade process works as expected in a containerized end to end test.
+
+We initialize the three primary Sourcegraph databases (frontend, codeintel-db, and codeinsights-db) from a historic version. Then use candidate builds of `frontend` and `migrator` to conduct a series of upgrades and validations. Validating expected state between steps.
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:14]:
```suggestion
We initialize the three primary Sourcegraph databases (frontend, codeintel-db, and codeinsights-db) from a historic version. Then use candidate builds of `frontend` and `migrator` to conduct a series of upgrades and validations, ensuring expected state between steps.
```

File: testing/tools/upgradetest/README.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+We conduct multiversion upgrades, and standard upgrades, based on their respective upgrade policies, only using MVU for versions in which it is necessary.
+
+- For Standard upgrades (migrator up) we test each patch version defined in the previous minor version of sourcegraph.
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:18]:
```suggestion
- For Standard upgrades (e.g. `migrator up`) we test each patch version defined in the previous minor version of sourcegraph.
```

File: testing/tools/upgradetest/README.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+- For MVU upgrades (migrator upgrade) we test all versions defined at least two minor versions prior to the latest patch release. i.e. all versions for which a standard upgrade will not work.
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:19]:
```suggestion
- For MVU upgrades (e.g. `migrator upgrade`) we test all versions defined at least two minor versions prior to the latest patch release. i.e. all versions for which a standard upgrade will not work.
```

File: testing/tools/upgradetest/README.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+- For autoupgrades we attempt an upgrade accross all versions.
+
+### Run Bazel Test
+
+```bash
+bazel test //testing/tools/upgradetest:sh_upgradetest --config darwin-docker -- <test args>
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:25]:
```suggestion
bazel test //testing/tools/upgradetest:sh_upgradetest -- <test args>
```

File: testing/tools/upgradetest/README.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+### Run Bazel Action
+
+- Version 0.0.0+dev (no autoupgrade):
+  ```bash
+  bazel run //testing/tools/upgradetest:sh_upgradetest_run --config darwin-docker -- <command>
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:32]:
```suggestion
  bazel run //testing/tools/upgradetest:sh_upgradetest_run -- <command>
```

File: testing/tools/upgradetest/README.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  ```
+- Stamp version:
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:34]:
```suggestion
- Stamped build:
```

File: testing/tools/upgradetest/README.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  ```bash
+  VERSION=x.x.x bazel run //testing/tools/upgradetest:sh_upgradetest_run --config darwin-docker --stamp --workspace_status_command=./dev/bazel_stamp_vars.sh -- <command>
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:36]:
```suggestion
  VERSION=x.x.x bazel run //testing/tools/upgradetest:sh_upgradetest_run --stamp --workspace_status_command=./dev/bazel_stamp_vars.sh -- <command>
```

File: testing/tools/upgradetest/README.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  ```
+
+### Run in CI
+
```
Comment (jhchabran) [inline testing/tools/upgradetest/README.md:40]:
```suggestion

Presently, the test runner is not plugged in CI, so the only way to get it to run is to trigger a custom build performing that specific test (i.e. a `bazel-do` CI runtype) 
```

File: testing/tools/upgradetest/test.sh [hunk 12]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+set -e
+
+RUNNER="$1"
+MIGRATOR_TARBALL="$2"
+FRONTEND_TARBALL="$3"
+
+# TODO why do we need this ? Is it hardcoded?
```
Comment (jhchabran) [inline testing/tools/upgradetest/test.sh:9]: cc @DaedalusG can we either clarify this or drop it?

File: testing/tools/upgradetest/README.md (hunk 2 above)
Comment (DaedalusG) [reply testing/tools/upgradetest/README.md:5]:
```
This test does not test Sourcegraph features, only the basic operations of database and frontend version upgrades and schema coherence.
```
^^ I've been starting to use schema coherence to describe the correctness of the schema definitions as well as the ability to traverse the migrators schema migration plan.

File: testing/tools/upgradetest/test.sh (hunk 12 above)
Comment (DaedalusG) [reply testing/tools/upgradetest/test.sh:9]: Dropping it, I don't remember why it was needed to begin with, and its commented out 🤷🏼‍♂️

File: testing/tools/upgradetest/README.md (hunk 1 above)
Comment (jhchabran) [reply testing/tools/upgradetest/README.md]: I think it's confusing as it introduces a new terminology that doesn't exist anywhere else. I'd rather say that you can use the tool on images that aren't properly tagged as long as they're built correctly.

**Review by jhchabran — APPROVED** [review]

**Pull request merged**